src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs

```diff
@@ -0,0 +1,55 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.WebSockets
+{
+    internal static class ManagedWebSocketExtensions
+    {
+        internal static string GetString(this UTF8Encoding encoding, Span<byte> bytes)
+        {
+            return encoding.GetString(bytes.ToArray());
+        }
+
+        internal static Task<int> ReadAsync(this Stream stream, Memory<byte> destination, CancellationToken cancellationToken)
+        {
+            if (destination.TryGetArray(out ArraySegment<byte> array))
+            {
+                return stream.ReadAsync(array.Array, array.Offset, array.Count, cancellationToken);
+            }
+            else
+            {
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(destination.Length);
+                return FinishReadAsync(stream.ReadAsync(buffer, 0, destination.Length, cancellationToken), buffer, destination);
+
+                async Task<int> FinishReadAsync(Task<int> readTask, byte[] localBuffer, Memory<byte> localDestination)
+                {
+                    try
+                    {
+                        int result = await readTask.ConfigureAwait(false);
+                        new Span<byte>(localBuffer, 0, result).CopyTo(localDestination.Span);
+                        return result;
+                    }
+                    finally
+                    {
+                        ArrayPool<byte>.Shared.Return(localBuffer);
+                    }
+                }
+            }
+        }
+    }
+
+    internal static class BitConverter
+    {
+        internal static int ToInt32(ReadOnlySpan<byte> value)
+        {
+            return System.BitConverter.ToInt32(value.ToArray(), 0);
```

Comment:
Same as earlier comment, this is very expensive; we should not allocate an array here.  You should be able to just copy the implementation from https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/BitConverter.cs#L268-L273.